CONTRIBUTING.md instructions (#707)
observablehq/plot: CONTRIBUTING.md

## Set up

To set up your local development environment for Plot, follow these steps on your terminal:

```bash
# 1. Clone your fork of the repository 
git clone https://github.com/USER-NAME/plot.git
cd plot

# 2. Install dependencies listed in the package.json file
yarn 
```

## Development

Plot is written in ES modules and uses [Vite](https://vitejs.dev/) for development; this means that you can edit the Plot source code, and the test plots will update live in your browser as you save changes. To start Vite:

```
yarn dev
```

This should open http://localhost:8008/ in your browser where you can see the tests plots defined in `test/plots`. 

![local development environment](img/localhost.png)

If you’d like to create a new test plot as part of your development, you need to:

1. Create a new `.js` file in `test/plots`, e.g., `test/plots/new-example.js` (an easy start is duplicating an existing example).

> 2. Export the function from your `new-example.js` file in `test/plots/index.js`. 

Reviewer: Isn’t this already covered below in the testing section, starting here?

> To add a new snapshot test, create a new JavaScript file in the `test/plots` folder.

Author: Good point -- as someone less familiar with this structure of development, I wouldn't think to look in the `test` section for information about how to create a new example (so I've left a sentence directing them there).

> Observable Plot is released under the ISC license. You are welcome to fork this repository and to send us pull requests to contribute bug fixes or new features.

Reviewer: ```suggestion
Observable Plot is released under the ISC license. You are welcome to [fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo) and to send us pull requests to contribute bug fixes or new features.
```

Observable Plot is released under the ISC license. You are welcome to fork this repository and to send us pull requests to contribute bug fixes or new features.

## Set up
To set up your local development environment for Plot, follow these steps on your terminal:

```bash
# 1. Clone this repository 
git clone git@github.com:observablehq/plot.git

> cd plot # cd into the folder

Reviewer: Does the comment “cd into the folder” help here, or does it just repeat what the code is saying? Is the expectation that the reader doesn’t know what “cd” does, and if so, should the comment explain that in English? It’s probably more appropriate to assume the reader knows what “cd” does, along with other terminal basics, here.

Author: Good point, we can remove the annotation.

# 2. Checkout a new development branch, by convention `user-name/feature-name`
git checkout -b user-name/feature-name

# 3. Install dependencies listed in the package.json file
yarn 
```

## Development
Plot is written in ES modules and uses [Vite](https://vitejs.dev/) for development; this means that you can edit the Plot source code and examples, and they’ll update live as you save changes. To start Vite:

```
yarn dev
```

This should open http://localhost:8008/ in your browser where you can see the tests plots defined in `test/plots`. 

> ![local deverlopment environment](img/localhost.png)

Reviewer 2: ```suggestion
![local development environment](img/localhost.png)
```

Reviewer 2: in img/localhost.png the names that are shown are the symbols exported by the js files in the test/plot/ folder. The file names are a bit different, and a few files (eg scales and legends) export multiple symbols.

> If you'd like to create a new example as part of your development, you need to:

Reviewer 2: ```suggestion
If you'd like to create a new test plot as part of your development, you need to:
```

these are not examples stricto sensu, but rather test plots that allow us to make sure that changes do not cause unwanted regressions in certain features. Examples are generally available in Observable notebooks, and (hopefully) convey best practices. Test plots are sometimes not good as examples.

1. Create a new `.js` file in `test/plots`, e.g., `test/plots/new-example.js` (an easy start is duplicating an existing example).

> 2. Export the function from your `new-example.js` file in `test/index.js`. Because the first alphabetical option is displayed by on the local server, you may want to export the example as `_newExample`.

Reviewer: I would not recommend this, as then folks are likely to submit a pull request with this naming convention. We want the tests to be in alphabetical order in `test/plots/index.js`. Also, once you’ve selected your test, you get hot module reloading, so you don’t have to keep re-selecting your test as you develop. And, you can use the left and right arrow keys to flip through the tests quickly, as well as Enter + typing to quickly select your desired test via the keyboard.

Reviewer: The test file is `test/plots/index.js`, not `test/index.js`.

Author: Noted about the naming convention (I hadn't used the hot-reloading enough to realize it kept your Select option). I'll remove that content.

Author: And I'll update the screenshot :)

### Development using Observable

> If you'd like to develop using Observable (or test your changes on an existing notebook), you'll neeed to first compile your changes into a single file:

Reviewer 2: ```suggestion
If you'd like to develop using Observable (or test your changes on an existing notebook), you'll neeed to first bundle your changes into a single file:
```

```bash
# Compile into a file `dist/plot.umd.js`
yarn prepublishOnly
```

You can then load the local copy of Plot in an Observable notebook:

```js
// Load your copy into Observable

> Plot = require("http://localhost:8008/path/to/plot/dist/plot.umd.js")

Reviewer: I don’t think we should recommend this approach specifically (because it doesn’t work in Safari), and more generally, I think we should not recommend that people “develop in Observable”. The recommended development pattern is to use Vite locally as described above. I think it’s okay (desirable, even) to describe how to _share a demo_ on Observable of some in-development feature. And in that case, you’d be uploading your built plot.umd.js as a file attachment and saying something like:

```js
Plot = require(await FileAttachment("plot.umd.js").url())
```

Author: I can appreciate that -- I'll remove this approach, and reframe the description about sharing on Observable.

rm -rf test/output
yarn test
```

## Submitting a pull request

> Once you have completed your feature and confirmed that all test have passed using `yarn test`, you should push your changes up to GitHub and submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

Reviewer: We mention pull requests up above (I suggest we move this GitHub link above), so maybe this section could instead be more specific about our expectations for what makes a good pull request, or a good feature? Like, please include a motivating example, or why you think this feature will be useful. Or, please consider filing an issue, or checking for existing issues, before submitting a pull request, in case someone else has already thought about this topic. Or perhaps to start, this section could instead be called `## Documentation`, and delete this sentence?

Author: I switched the last section to **Documentation** and moved the Observable instructions there.

> Please make sure to update the `README.md` file to reflect changes to the library.

Reviewer: ```suggestion
Please update the [README.md](./README.md) to reflect any changes to Plot’s public API.
```

rm -rf test/output
yarn test
```
## Submitting a pull request
Once you have completed your feature and confirmed that all test have passed using `yarn test`, you should push your changes up to GitHub and submit a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request).

```bash
# Push changes up to a new branch on GitHub

> git push -u origin user-name/some-feature

Reviewer 2: ```suggestion
git push -u origin user-name/some-feature
```

I don't think this workflow is generally available. You should first fork the repo, then push to a new branch on your fork, and finally make a pull-request against the official repo.

Author: This is interesting -- I wonder what a good way to callout different processes for our colleagues might be...

Reviewer: I’m not sure it’s appropriate to cover the basics of git, GitHub, terminal commands here. It might be better to link to the corresponding GitHub tutorials, e.g.,

https://docs.github.com/en/get-started/quickstart/fork-a-repo
https://docs.github.com/en/get-started/quickstart/contributing-to-projects
